### ISSUE TITLE
Invalid test cases for runtime.Marshaler

### ISSUE THREAD

**Issue by movsb** [grpc-ecosystem/grpc-gateway#1501]
## 🐛 Bug Report

Taking address of any element in this slice produces the same value (because the element is an empty struct):

https://github.com/grpc-ecosystem/grpc-gateway/blob/855204b63e5a0977e67eeedfecb64f6fec7228cf/runtime/marshaler_registry_test.go#L30-L52

So, these test cases almost won't work as expected.

As per [the last line of golang spec](https://golang.org/ref/spec#Size_and_alignment_guarantees):

> A struct or array type has size zero if it contains no fields (or elements, respectively) that have a size greater than zero.
> 
> Two distinct zero-size variables may have the same address in memory. 

## To Reproduce

See the code at [playground](https://play.golang.org/p/HPwRXhccsU9):

```go
package main

import (
	"fmt"
)

type S struct {
}

type T struct {
	b bool
}

func main() {
	var s [3]S
	var t [3]T

	fmt.Printf("%p,%p,%p\n", &s[0], &s[1], &s[2])
	fmt.Printf("%p,%p,%p\n", &t[0], &t[1], &t[2])
}
```

Outputs:

```-
0x58fd18,0x58fd18,0x58fd18
0xc00009400b,0xc00009400c,0xc00009400d
```

## Your Environment

Any.

**Comment by johanbrandhorst** [comment]
Hi, thanks for the bug report! Do you have any ideas for how we could fix this? I think this is a pretty old part of the code.

**Comment by yugui** [comment]
Thank you for reporting this bug, @movsb. That was my fault.
Fortunately the actual implementation of marshaler registry was implemented as expected.

> Do you have any ideas for how we could fix this? I think this is a pretty old part of the code.

Giving a non-zero size to the type could solve the issue. (#1502).
I found that the test spec I wrote was completely wrong and that the zero-length type has been hiding it.  So I also corrected the spec.
